### PR TITLE
Create ROI (when there is no RTSS)

### PR DESCRIPTION
**Pull request by DuyTungHa**
Please stress test the program (try to think of edge cases). 
From now on, there will only be one instance of DrawROIWindow and DeleteROIWindow in the program to avoid memory leak issues. 
ISO2ROI functionality will no longer automatically save RTSS because of the possibility of merging conflicts.

### CODE REVIEW

File: test/test_model_roi.py [hunk 1]
```diff
@@ -94,4 +94,3 @@ def test_create_roi():
     )
     assert (first_contour.ContourGeometricType == "CLOSED_PLANAR")
     assert (rt_ss.RTROIObservationsSequence[0].RTROIInterpretedType == "ORGAN")
-
```
Comment (tdlinh2712) [inline test/test_model_roi.py:97]: please add unit test for generate_rtss_from_ct function

Comment (DuyTungHa) [reply test/test_model_roi.py:97]: any ideas how to test this function ?

File: src/View/ImageLoader.py [hunk 2]
```diff
@@ -123,8 +128,32 @@ def load(self, interrupt_flag, progress_callback):
...
+            # Add RT Struct file path and dataset to patient dict container
+            patient_dict_container.filepaths['rtss'] = rtss_path
+            patient_dict_container.dataset['rtss'] = rtss
+
+            # Set some patient dict container attributes
+            patient_dict_container.set("file_rtss", rtss_path)
+            patient_dict_container.set("dataset_rtss", rtss)
+            ordered_dict = DicomTree(None).dataset_to_dict(rtss)
+            patient_dict_container.set("dict_dicom_tree_rtss", ordered_dict)
+            patient_dict_container.set("selected_rois", [])
```
Comment (tdlinh2712) [inline src/View/ImageLoader.py]: this load function (from the old code) is already very long and contains nested if-else, which will make it very difficult to maintain and expand. Can we refactor it and divide into smaller functions? or at least, put the added code in a sepeare function and call it here.

File: src/View/OpenPatientWindow.py [hunk 3]
```diff
@@ -324,6 +325,14 @@ def tree_item_clicked(self, item, _):
             header = "Only one study can be opened."
             self.open_patient_window_confirm_button.setDisabled(True)
 
+        # If RTSTRUCT exists but not selected, save existing RTSTRUCT file path
+        self.existing_rtss_path = None
+        if 'RTSTRUCT' not in selected_series_types and len(selected_series_types) > 0:
+            selected_study = self.get_checked_leaves()[0].parent()
```
Comment (tdlinh2712) [inline src/View/OpenPatientWindow.py:340]: put this in a separate function, save_existng_rtss or something

Comment (tdlinh2712) [reply src/View/OpenPatientWindow.py:340]: do something like self.existing_rtss_path = self.get_existing_rtss()
then in `get_existing_rtss(self)` return None or appropriate path

File: src/View/mainpage/StructureTab.py [hunk 4]
```diff
@@ -343,7 +367,100 @@ def save_new_rtss(self, event=None):
...
+            button_yes = QtWidgets.QPushButton("Yes, I want to merge")
+            button_no = QtWidgets.QPushButton("No, I will change the names")
+            """ We want the buttons 'No' & 'Yes' to be displayed in that exact order. QMessageBox displays buttons in
+                respect to their assigned roles. (0 first, then 0 and so on) 'AcceptRole' is 0 and 'RejectRole' is 1 
+                thus by counterintuitively assigning 'No' to 'AcceptRole' and 'Yes' to 'RejectRole' the buttons are 
+                positioned as desired.
+            """
+            confirm_merge.addButton(button_no, QtWidgets.QMessageBox.AcceptRole)
+            confirm_merge.addButton(button_yes, QtWidgets.QMessageBox.RejectRole)
+            confirm_merge.exec_()
```
Comment (tdlinh2712) [inline src/View/mainpage/StructureTab.py]: put this in a function `display_dupicate_names_warning` or something similar

File: src/View/mainpage/StructureTab.py [hunk 5]
```diff
@@ -343,7 +367,100 @@ def save_new_rtss(self, event=None):
...
+
+        # Renumber the ROINumber and ReferencedROINumber tags
+        original_structure_set = StructureTab.renumber_roi_number(original_structure_set)
+        original_roi_contour = StructureTab.renumber_roi_number(original_roi_contour)
+        original_roi_observation_sequence = StructureTab.renumber_roi_number(original_roi_observation_sequence)
+
+        # Set the new value
+        old_rtss.add_new(Tag("StructureSetROISequence"), "SQ",original_structure_set)
+        old_rtss.add_new(Tag("ROIContourSequence"), "SQ", original_roi_contour)
+        old_rtss.add_new(Tag("RTROIObservationsSequence"), "SQ", original_roi_observation_sequence)
```
Comment (tdlinh2712) [inline src/View/mainpage/StructureTab.py]: this looks like Model logic? can we put it in a function in Model?

File: src/View/mainpage/StructureTab.py [hunk 6]
```diff
@@ -343,7 +367,100 @@ def save_new_rtss(self, event=None):
...
+        return old_rtss
+
+    @staticmethod
+    def renumber_roi_number(sequence):
+        roi_number = 1
+        for item in sequence:
+            if item.get("ROINumber") is not None:
+                item.add_new(Tag("ROINumber"), 'IS', str(roi_number))
+            elif item.get("ReferencedROINumber") is not None:
+                item.add_new(Tag("ReferencedROINumber"), "IS", str(roi_number))
```
Comment (tdlinh2712) [inline src/View/mainpage/StructureTab.py]: this one as well. What does it do?

**Review by tdlinh2712 — CHANGES_REQUESTED** [review]

File: src/View/ImageLoader.py (hunk 2 above)
Comment (DuyTungHa) [reply src/View/ImageLoader.py]: Done

File: src/View/OpenPatientWindow.py (hunk 3 above)
Comment (DuyTungHa) [reply src/View/OpenPatientWindow.py:340]: Done

File: src/View/mainpage/StructureTab.py (hunk 4 above)
Comment (DuyTungHa) [reply src/View/mainpage/StructureTab.py]: I don't think that we should make this change because in other places in the program, the confirm box code is put inside the function. Separating code into too many small functions will harm the readability of the code.

File: src/View/mainpage/StructureTab.py (hunk 5 above)
Comment (DuyTungHa) [reply src/View/mainpage/StructureTab.py]: It looks like Model logic but it is solely used in Structure Tab to save rtss so I don't think it is a good idea to put this in a model class. Future developers may have difficulty in finding where stuff are located.

File: src/View/mainpage/StructureTab.py (hunk 6 above)
Comment (DuyTungHa) [reply src/View/mainpage/StructureTab.py]: This is used to renumber the ROI in in the merged RTSS. I don't recommend putting it elsewhere because it is highly related to merge rtss functionality.

File: src/View/mainpage/StructureTab.py (hunk 5 above)
Comment (tdlinh2712) [reply src/View/mainpage/StructureTab.py]: @DuyTungHa they can find it by right click "go to Declaration of Implementation". I recommend putting it in a Model folder, maybe add to Model/ROI.py or create a file Model/RTSS.py to store the function. (Line413 -> end of this function). Putting the logic in a View class diminishes the point of MVC and the role of Model, View and Controller components. I know this is a current theme in Onko but we better not add to that

Comment (DuyTungHa) [reply src/View/mainpage/StructureTab.py]: Done

File: test/test_model_roi.py (hunk 1 above)
Comment (DuyTungHa) [reply test/test_model_roi.py:97]: Done
